File: src/main/java/com/pauljoda/modularsystems/storage/gui/GuiStorageCore.java
package com.pauljoda.modularsystems.storage.gui;

import com.pauljoda.modularsystems.core.ModularSystems;
import com.pauljoda.modularsystems.storage.container.ContainerStorageCore;
import com.pauljoda.modularsystems.storage.tiles.TileStorageCore;
import com.teambr.bookshelf.client.gui.GuiBase;
import com.teambr.bookshelf.client.gui.component.BaseComponent;
import com.teambr.bookshelf.client.gui.component.control.GuiComponentButton;
import com.teambr.bookshelf.client.gui.component.control.GuiComponentScrollBar;
import com.teambr.bookshelf.client.gui.component.control.GuiComponentTextBox;
import com.teambr.bookshelf.client.gui.component.control.GuiComponentTexturedButton;
import com.teambr.bookshelf.client.gui.component.display.GuiComponentText;
import com.teambr.bookshelf.client.gui.component.display.GuiTabCollection;
import com.teambr.bookshelf.helpers.GuiHelper;
import com.teambr.bookshelf.manager.PacketManager;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.util.StatCollector;
import org.lwjgl.input.Keyboard;
import org.lwjgl.input.Mouse;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Modular-Systems
 * Created by Paul Davis on 7/23/2015
 */
public class GuiStorageCore extends GuiBase<ContainerStorageCore> {
    private boolean isInMainArea;
    private float currentScroll;

    public TileStorageCore core;

    private GuiComponentTextBox textBox;
    private GuiComponentScrollBar scrollBar;

    /**
     * Constructor for All Guis
     *
     * @param container Inventory Container
     */
    public GuiStorageCore(ContainerStorageCore container, TileStorageCore tile) {
        super(container, 250, 220, tile.getInventoryName());
        title.setXPos(8);
        core = tile;
        addRightTabs(rightTabs);
        addComponents();
    }

    @Override
    public void addComponents() {
        if(core != null) {
            components.add(textBox = new GuiComponentTextBox(95, 5, 150, 16) {
                @Override
                public void fieldUpdated(String value) {
                    inventory.keyTyped(value);
                }
            });

            textBox.getTextField().setVisible(core.hasSearchUpgrade());

            components.add(scrollBar = new GuiComponentScrollBar(227, 26, 108) {
                @Override
                public void onScroll(float position) {
                    inventory.scrollTo(position);
                    currentScroll = position;
                }
            });

            if(core.hasSortingUpgrade()) {
                components.add(new GuiComponentButton(3, 26, 20, 20, "S") {
                    @Override
                    public void doAction() {
                        core.sortInventory();
                        PacketManager.updateTileWithClientInfo(core);
                    }

                    @Override
                    public List<String> getDynamicToolTip(int mouseX, int mouseY) {
                        return Collections.singletonList(StatCollector.translateToLocal("inventory.storageCore.sort"));
                    }

                });
            }

            if(core.hasCraftingUpgrade()) {
                components.add(new GuiComponentTexturedButton(228, 147, 76, 247, 9, 9, 11, 11) {
                    @Override
                    public void doAction() {
                        clearCraftingGrid();
                    }

                    @Override
                    public List<String> getDynamicToolTip(int mouseX, int mouseY) {
                        return Collections.singletonList(StatCollector.translateToLocal("inventory.storageCore.clear"));
                    }
                });
            }
        }
    }

    public void clearCraftingGrid() {
        inventory.clearCraftingGrid();
        PacketManager.updateTileWithClientInfo(core);
        for(int i = 0; i < this.mc.thePlayer.inventory.getSizeInventory(); i++) {
            this.mc.playerController.sendSlotPacket(this.mc.thePlayer.inventory.getStackInSlot(i), i);
        }
    }

    @Override
    protected void keyTyped(char letter, int keyCode) {
        if(textBox != null && textBox.getTextField() != null && !textBox.getTextField().textboxKeyTyped(letter, keyCode)) {
            super.keyTyped(letter, keyCode);
        } else if(textBox != null && textBox.getTextField() != null){
            this.inventory.keyTyped(textBox.getValue());
        } else
            super.keyTyped(letter, keyCode);
    }

    protected void mouseClicked(int par1, int par2, int par3) {
        super.mouseClicked(par1, par2, par3);
        if(par3 == 2 && core != null && core.hasSortingUpgrade()) {
            core.sortInventory();
            PacketManager.updateTileWithClientInfo(core);
            GuiHelper.playButtonSound();
        }
    }

    @Override
    public void handleMouseInput() {
        super.handleMouseInput();
        int scrollDirection = Mouse.getEventDWheel();
        if(scrollDirection != 0 && inventory.storage.getInventoryRowCount() > 6) {
            //Since NEI messes with stuff, lets only scroll if its not there or we are outside the main area
            if(ModularSystems.nei != null && isInMainArea)
                return;

            int rowsAboveDisplay = this.inventory.storage.getInventoryRowCount() - 6;
            if(scrollDirection > 0)
                scrollDirection = 1;
            else
                scrollDirection = -1;

            this.currentScroll = (float)((double)this.currentScroll - (double)scrollDirection / (double)rowsAboveDisplay);

            if(currentScroll < 0.0F)
                currentScroll = 0.0F;
            else if(currentScroll > 1.0F)
                currentScroll = 1.0F;

            this.inventory.scrollTo(this.currentScroll);
            this.scrollBar.setPosition(this.currentScroll);
            updateScreen();
        }
        if(textBox != null && textBox.getTextField() != null && textBox.getTextField().isFocused())
            Keyboard.enableRepeatEvents(true);
        else
            Keyboard.enableRepeatEvents(false);
    }

    @Override
    public void drawScreen(int x, int y, float f) {
        super.drawScreen(x, y, f);
        isInMainArea = GuiHelper.isInBounds(x, y, guiLeft + 35, guiTop + 17, guiLeft + 220, guiTop + 215);
    }

    /**
     * Adds the tabs to the right. Overwrite this if you want tabs on your GUI
     * @param tabs List of tabs, put GuiTabs in
     */
    @Override
    public void addRightTabs(GuiTabCollection tabs) {
        if(core != null) {
            List<BaseComponent> children = new ArrayList<>();
            children.add(new GuiComponentText("Information", 26, 6, 0xFFCC00));
            children.add(new GuiComponentText("Size: ", 10, 23, 0xFFFFFF));
            children.add(new GuiComponentText(String.valueOf(core.getSizeInventory()) + " Slots", 20, 33, 0x777777));
            tabs.addTab(children, 100, 100, new Color(77, 75, 196), new ItemStack(Items.book, 1));
        }
    }
}

File: src/main/scala/com/teambr/modularsystems/storage/gui/GuiStorageCore.scala
package com.teambr.modularsystems.storage.gui

import java.awt.Color

import com.teambr.bookshelf.client.gui.{GuiTextFormat, GuiColor, GuiBase}
import com.teambr.bookshelf.client.gui.component.BaseComponent
import com.teambr.bookshelf.client.gui.component.control.{GuiComponentTextBox, GuiComponentScrollBar}
import com.teambr.bookshelf.client.gui.component.display.{GuiComponentText, GuiTabCollection}
import com.teambr.bookshelf.helper.LogHelper
import com.teambr.modularsystems.core.network.PacketManager
import com.teambr.modularsystems.core.utils.ClientUtils
import com.teambr.modularsystems.storage.container.ContainerStorageCore
import com.teambr.modularsystems.storage.network.{UpdateFilterString, ScrollStorageCore}
import com.teambr.modularsystems.storage.tiles.TileStorageCore
import net.minecraft.client.Minecraft
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.item.ItemStack
import net.minecraft.util.text.translation.I18n
import org.lwjgl.input.{Keyboard, Mouse}

import scala.collection.mutable.ArrayBuffer

/**
  * This file was created for Modular-Systems
  *
  * Modular-Systems is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis "pauljoda"
  * @since 3/27/2016
  */
class GuiStorageCore(player: EntityPlayer, tile: TileStorageCore)
        extends GuiBase[ContainerStorageCore](new ContainerStorageCore(player.inventory, tile),
            250, 240, "inventory.storage.title") {

    if(tile.hasSearchUpgrade)
        components(0).setXPos(10)

    lazy val ourRender = new RenderItemLarge(Minecraft.getMinecraft.getTextureManager,
        Minecraft.getMinecraft.getRenderItem.getItemModelMesher.getModelManager,
        Minecraft.getMinecraft.getItemColors)

    var currentScroll = 0.0F

    var updateBar = false
    var TEXT_BOX : Int = 2

    override def setWorldAndResolution(mc: Minecraft, width: Int, height: Int): Unit = {
        super.setWorldAndResolution(mc, width, height)
        itemRender = ourRender
    }

    override def updateScreen(): Unit = {
        inventory.detectAndSendChanges()
        super.updateScreen()
    }

    override def handleMouseInput(): Unit = {
        super.handleMouseInput()
        var scrollDirection = Mouse.getEventDWheel

        if(scrollDirection != 0 && inventory.storageCore.getInventory.size() / 11 >= inventory.rowCount) {

            val rowsAboveDisplay = (inventory.storageCore.getInventory.size() / 11) - inventory.rowCount
            if(scrollDirection > 0)
                scrollDirection = 1
            else
                scrollDirection = -1

            currentScroll = (this.currentScroll.toDouble - (scrollDirection.toDouble / rowsAboveDisplay.toDouble)).toFloat

            if(currentScroll < 0.0F)
                currentScroll = 0.0F
            else if(currentScroll > 1.0F)
                currentScroll = 1.0F

            this.inventory.scrollTo(currentScroll)
            PacketManager.net.sendToServer(new ScrollStorageCore(currentScroll))
            updateBar = true
        }

        components(TEXT_BOX) match {
            case textBox : GuiComponentTextBox =>
                if (textBox != null && textBox.getTextField != null && textBox.getTextField.isFocused)
                    Keyboard.enableRepeatEvents (true)
                else
                    Keyboard.enableRepeatEvents (false)
            case _ => LogHelper.debug("This is not the text box")
        }
    }

    override def keyTyped(typedChar: Char, keyCode: Int): Unit = {
        if (keyCode == this.mc.gameSettings.keyBindInventory.getKeyCode) {
            if(!Keyboard.areRepeatEventsEnabled())
                super.keyTyped(typedChar, keyCode)
            else {
                components(TEXT_BOX) match {
                    case textBox : GuiComponentTextBox =>
                        textBox.keyTyped(typedChar, keyCode)
                    case _ => LogHelper.debug("This is not the text box")
                }
            }
        } else
            super.keyTyped(typedChar, keyCode)
    }

    override def addComponents(): Unit = {
        // Scroll Bar
        components += new GuiComponentScrollBar(227, 26, 18 * inventory.rowCount) {
            override def onScroll(position: Float): Unit = {
                inventory.scrollTo(position)
                PacketManager.net.sendToServer(new ScrollStorageCore(currentScroll))
                currentScroll = position
            }

            override def render(guiLeft: Int, guiTop: Int, mouseX : Int, mouseY : Int): Unit = {
                if(updateBar) {
                    setPosition(currentScroll)
                    updateBar = false
                }
                super.render(guiLeft, guiTop, mouseX, mouseY)
            }
        }

        // Search Bar
        if(tile != null && tile.hasSearchUpgrade) {
            components += new GuiComponentTextBox(132, 7, 108, 16) {
                override def fieldUpdated(value: String): Unit = {
                    inventory.filterString = value
                    PacketManager.net.sendToServer(new UpdateFilterString(value))
                }
            }
        }
    }

    /**
      * Adds the tabs to the right. Overwrite this if you want tabs on your GUI
      *
      * @param tabs List of tabs, put GuiTabs in
      */
    override def addRightTabs(tabs: GuiTabCollection) : Unit = {
        if(tile != null) {
            var infoTab = new ArrayBuffer[BaseComponent]()
            // Information
            infoTab += new GuiComponentText(GuiColor.ORANGE  + I18n.translateToLocal("inventory.info"), 26, 6)

            // Quantity
            infoTab += new GuiComponentText(
                GuiColor.YELLOW + "" + GuiTextFormat.BOLD + I18n.translateToLocal("inventory.storageCore.quantity"),
                5, 23)
            infoTab += new GuiComponentText(GuiColor.WHITE + ClientUtils.formatNumber(tile.getCachedSize), 15, 33) {
                override def render(x : Int, y : Int, mx : Int, my : Int): Unit = {
                    setText(GuiColor.WHITE + ClientUtils.formatNumber(tile.getCachedSize))
                    super.render(x, y, mx, my)
                }
            }

            // Capacity
            infoTab += new GuiComponentText(
                GuiColor.YELLOW + "" + GuiTextFormat.BOLD + I18n.translateToLocal("inventory.storageCore.capacity"),
                5, 48)
            infoTab += new GuiComponentText(GuiColor.WHITE + ClientUtils.formatNumber(tile.getMaxItems), 15, 58) {
                override def render(x : Int, y : Int, mx : Int, my : Int): Unit = {
                    setText(GuiColor.WHITE + ClientUtils.formatNumber(tile.getMaxItems))
                    super.render(x, y, mx, my)
                }
            }

            tabs.addTab(infoTab.toList, 100, 100, new Color(150, 112, 50), new ItemStack(Items.book))
        }
    }
}
